File: 01-what-is-cpp/1-memory-cpp-vs-java/Memory.java

/**
 * Demonstrates that memory usage in Java is surprising.
 */

class Point {
	private int x;
	private int y;
	public Point() { x = y = 5555; }	
};


public class Memory {
	static final int KB=1024;
	static final int SIZE=50*1000*KB;
 
	public static void main(String[] args) throws InterruptedException {
		System.out.println("Before new");
		Thread.sleep(3000);

		Point[] p = new Point[SIZE];
		for (int  i=0; i<SIZE; ++i)
			p[i] = new Point();

		System.out.println("After new");

		// The free memory should drop by ... KB

		Thread.sleep(10000);
	}
}

File: 01-what-is-cpp/1-memory-cpp-vs-java/Memory.cpp
/**
 * Demonstrates memory usage in C++.
 * @author erelsgl
 */

#include <iostream>
#include <thread>
#include <chrono>
using namespace std;

class Point {
	private:
		int x;
		int y;
	public:
		Point() { x = y = 5555; }	
};

int main() {
	const int KB = 1024;
	const int SIZE=50*1000*KB;

	cout << "Before new" << ' ' << endl;
	this_thread::sleep_for(chrono::seconds(3));

	Point* p = new Point[SIZE];
	for (int  i=0; i<SIZE; ++i)
		p[i] = Point();
		
	cout << "After new" << endl;
	// The free memory should drop by ... KB.

	this_thread::sleep_for(chrono::seconds(10));

	delete[] p;
}
